vllm/v1/core/single_type_kv_cache_manager.py

```diff
@@ -231,19 +231,48 @@ def find_longest_cache_hit(
 
         raise NotImplementedError
 
-    @abstractmethod
     def remove_skipped_blocks(self, request_id: str,
                               num_computed_tokens: int) -> None:
         """
         Remove the blocks that are no longer needed from `blocks` and free the 
         blocks. The removed blocks should be replaced by null_block.
         Need to be customized for each attention type.
-
         Args:
             request_id: The request ID.
             num_computed_tokens: The number of tokens that have been computed.
         """
-        raise NotImplementedError
+        # Remove the blocks that are no longer be in the sliding window and
+        # skipped during the attention computation.
+        last_useful_token = self.get_last_useful_token(num_computed_tokens)
+        if last_useful_token <= 0:
+            # This indicates that ALL tokens are inside attention window.
+            # Thus we do not need to free any blocks outside attention window.
+            return
+        last_useful_block = last_useful_token // self.block_size
+        blocks = self.req_to_blocks[request_id]
+        removed_blocks: list[KVCacheBlock] = []
+        for i in range(last_useful_block - 1, -1, -1):
+            if blocks[i] == self._null_block:
+                # If the block is already a null block, the blocks before it
+                # should also have been set to null blocks by the previous calls
+                # to this function.
+                break
+            removed_blocks.append(blocks[i])
+            blocks[i] = self._null_block
+        self.block_pool.free_blocks(removed_blocks)
+
+    def get_last_useful_token(self, num_computed_tokens: int) -> int:
```

Comment:
Prefer `get_num_skipped_tokens`. `get_last_useful_token` is a bit confusing as people have to know it is 0-indexed.